---
title: "Locus-Based Restructure"
author: "Nick O'Brien"
date: "26/06/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

The past few weeks I've been working on restructuring my genome to run faster, and be more realistic. I've also been reading about the natural variability in my parameters to inform the choices of the ranges I'm using for my latin hypercube samples. 
I also started using Tinaroo, moving from Awoonga. This allows for way more cores, with pretty much the same system I used for parallelism as before: my only change was using the function:

``` {r cluster, eval=FALSE}
makeCluster(future::availableWorkers(methods = "PBS"))
```
instead of availableCores(), to utilise multiple nodes (and multiple cores on each node) at once.

## Locus-based model

My model's length is now considerably shorter, from 200,000bp to ~20,000 loci. Since SLiM's genomes are quite arbitrary, each element of a genome vector is not explicitly a 'base-pair', but just an element of a vector. Hence, it is possible to treat each of these elements as a locus rather than a base pair, shortening the length of the genome considerably. There are a few assumptions associated with this. First of all, there is an implied length of base pairs within each of the locus 'boxes' along the genome, which is not explicitly modelled. The length of each locus in terms of the number of these pseudo-basepairs is assumed to be the same. This is reasonably accurate, as eukaryotic (and prokaryotic!) protein-coding gene lengths tend to be conserved, as noted by Xu et al. 2006. Their estimate of an average gene size was 1346bp, which I am using as the base pair 'length' of each of my loci(more on that later). The next assumption is that each of these implictly-modelled base pairs has an equal chance of having a mutation appear there, and all the mutations have the same effect (equivalent). While I have not been able to confirm/deny this yet from the literature, I believe it is reasonable to assume, as is that each base pair is independent. A mutation at the same position is stacked with the others, because it is assumed that this mutation is more likely to occur at a different base-pair within the locus, than outside of it, or be more complex than a point-mutation. This system allows for a more grounded approach by eliminating the assumption that all phenotype-affecting mutations are point mutations, and instead saying they are any mutation at any point within the locus. This is also more realistic in terms of real data, where the exact point of a mutation within a locus may not be known, but its effects are.

Since mutation rates and recombination rates in SLiM's documentation are described in chance per base pair per generation, we need to be clear that with this model it is actually chance per locus per generation, and be very careful we adjust for that when selecting parameters. The probability of a mutation occurring at a given locus is actually the sum of the probabilities of that mutation occurring at each base pair within the locus (which is assumed to be equal). So if we find a recombination rate expressed in per base pair units, for example, 1e-8, we need to adjust be the average number of base pairs per locus, which I have already mentioned at 1346bp, resulting in a per locus rate of 1.346e-5. Some genes may have higher or lower mutation and recombination rates (and bp lengths), for a purely general standpoint, this average should do, perhaps with some fluctuation around that from a normal distribution. To improve this, I will read into what distributions best explain genome-wide recombination rate to find how to best implement varying rates (same with mutation rates).

```{Rcpp restructure1, eval=FALSE}
initialize() {
	setCfgParam("mu", 6.3e-6); // Genome-wide mutation rate per locus, Schug et al. 1996, as above
	setCfgParam("Ne", 10000); // Population size
	setCfgParam("rregion", 0.5); // Recombination rate between genomic regions
	setCfgParam("rwide", 1.346e-5); // Recombination rate across the whole genome (aside from the between genomic regions rate)
	setCfgParam("genomelength", 20680); // Taken from my own analysis from Ensembl, and from Elliott and Gregory 2015 
	
	
	// g1 genomic element type: uses m1 and m2 for all mutations - get other effects only if determined for m1 
	initializeGenomicElementType("g1", c(m1, m2), c(1, delmu));
	
	
	// Chromosome set up with 10 genomic regions, where there are 200 potential QTL sites, the rest are neutral sites (m1, m2)
	initializeGenomicElement(g1, 0, genomelength - 1);
	
	recomseq = seq(0, genomelength - 1, by = 0.1*genomelength);
	ends = asInteger(round(recomseq[1:9])); 	// Convert to int from float
	ends = c(sort(c(ends, ends+1)), genomelength);
	rbtwnQTL = c(rep(rregion, length(ends)));
	rgenome = c(rep(rwide, length(ends))); 
	rates = c(rbind(rgenome, rbtwnQTL), rwide);
	rates = rates[0:(length(ends) - 1)];
	
	initializeRecombinationRate(rates, ends); // We set this later, when we get the positions of QTLs
	
	
}
```

## Direct control over number of loci

I also implemented a system for controlling how many loci are affecting a trait. This is done by removing natural creation of any QTL mutations, and instead determining their creation through a mutation callback. Only m1 (neutral, Null mutations) and m2 (deleterious, non-trait) mutations can appear. When an m1 mutation appears, the code checks if that mutation is occurring at a pre-determined trait locus (which is stored in the simulation at the first generation). If it is, the mutation is changed to a QTL mutation of a certain type according to a binomial flip with probability equal to the rate of pleiotropic mutations. If the flip returns true, the mutation becomes m3 (pleiotropic), and its effects on both traits are pulled as before. Otherwise, the mutation type is chosen at random (via a discrete uniform distribution) from the other QTL mutation types (m4 and m5; affecting either trait 0 or trait 1). This is quite easily expanded when I complete my four-trait model. The chance that a mutation at a QTL position remains synonymous/null is included within the distribution of effects: a normal distribution, N(0, 1). since the mean is 0, on average the mutations are null, or close to it. Since we have direct control over where QTL mutations can appear, this system allows for direct control over linkage via changing the distribution from which the initial QTLs are pulled from - at the moment I am just using sample() to randomly choose numbers from across the genome, but I could easily use rnorm(), runif(), rexp(), or rgamma(), and that is just within SLiM - I could also feed a vector of values in externally, using any number of distributions or hand-selected locus positions from R.


```{Rcpp restructure2, eval=FALSE}
initialize() {

	setCfgParam("nloci", 100); // Number of QTL positions along the chromosome - where QTL mutations can appear, sampled from sample()
	
	initializeMutationRate(mu);
	
	initializeMutationType("m1", 0.5, "f", 0.0); // Neutral, non-trait-affecting mutations: remove these?
	
	//m2 mutation type: deleterious/background selection
	initializeMutationType("m2", 0.5, "g", del_mean, del_shape);  // del_mean and del_shape are defined in the batch run script, or the rwide script
	m2.color = "red";
	
	//m3 mutation type: pleiotropic QTL for both traits, fixed, we pull effects from rmvnorm
	initializeMutationType("m3", 0.5, "f", 0.0);
	//	m3.convertToSubstitution = F;
	m3.color = "green";
	
	//m4 mutation type: QTL for phenotype 0 only
	initializeMutationType("m4", 0.5, "n", 0.0, 1.0);
	//	m4.convertToSubstitution = F;
	m4.color = "blue";
	
	//m5 mutation type: QTL for phenotype 1 only
	initializeMutationType("m5", 0.5, "n", 0.0, 1.0);
	//	m5.convertToSubstitution = F;
	m5.color = "gray";

	// QTL constants
	
	setCfgParam("QTL_mu", c(0, 0)); // mean distribution value for multivariate normal we are pulling effects of both traits from
	setCfgParam("QTL_cov", 0.0); // covariance between the pulls in the multivariate normal (pleiotropic correlation between traits)
	setCfgParam("QTL_sigma", matrix(c(1, QTL_cov, QTL_cov, 1), nrow=2)); // sigma used in multivariate normal pull - variance covariance matrix, where variance of each trait = 1


	
// create a population of Ne individuals
1 {
	sim.addSubpop("p1", Ne);
	chr_QTL = sort(sample(1:(sim.chromosome.lastPosition-1), nloci));

	sim.setValue("chr_QTL", chr_QTL); //store these positions in a sim value
	
}

// Set fitness of m4 and m5 to be neutral
fitness(m4) {
	return 1.0;
}

fitness(m5) {
	return 1.0;
}

// Check if the m1 mutation's position is at a chr_QTL position, if it is than redraw the mutationType

mutation(m1) {
	if (any(mut.position == sim.getValue("chr_QTL"))) { // If the mutation falls in any of our defined QTL loci within a region, transform it to pleiotropic, or non-pleiotropic QTL - the chance of the QTL still being null/synonymous is implicit in the mean QTL value being 0
		
		pleio = rbinom(1, 1, pleiorate); // Redraw the mutation type, determining whether the nonsyn mutation should be pleiotropic or not 	
		
		if (pleio == 1) {
			mut.setMutationType(m3);
			effects = rmvnorm(1, QTL_mu, QTL_sigma);		// Calculate effects for m3 here since a mutation callback won't be called since it isn't autogenerated
			mut.setValue("e0", effects[0]);
			mut.setValue("e1", effects[1]);
			mut.setValue("fe", 0.0); //fitness effect
		}
		else {
			type = rdunif(1, 3, (length(sim.mutationTypes) - 1)); // If it isn't pleiotropic, randomly choose one of the phenotypes for the QTL to affect
			mut.setMutationType(sim.mutationTypes[type]);
			mut.setSelectionCoeff(mut.mutationType.drawSelectionCoefficient(1));
			
			if (mut.mutationType == m4)	{
				mut.setValue("e0", mut.selectionCoeff);				// Calculate effects for each, since it isn't 
				mut.setValue("e1", 0.0);
				mut.setValue("fe", 0.0); //fitness effect
			}
			
			if (mut.mutationType == m5) {
				mut.setValue("e0", 0.0);
				mut.setValue("e1", mut.selectionCoeff);
				mut.setValue("fe", 0.0); //fitness effect	
			}
		}
		return T;
	}
	else
		
		return T;
}

}
```

The fitness effect value there is for standardising output: since m2 mutations have an effect on fitness, I wanted to include that in my mutational output. However, I will need to rethink this for beneficial mutations, as there it isn't the mutation values that are directly affecting fitness, but the sum of them all to a phenotype, which is either close to or far away from a phenotype optimum, and that total phenotype determines fitness.


## Matrix outputs

I've also started discussing output for the mathematicians in the new centre Daniel is starting - this has led me to define three new tables/matrices that together describe everything in my standardised trait mean/variance output, but in a way that graph theory can be used. These matrices are either individuals * nloci, or genomes * nloci, depending on the table; segregating sites and size effects are genomes, zygosity is individuals.
At the moment this runs pretty slowly (I think it is O(n^2)), but I have an idea of how to improve it via sapply() rather than looping through each individual/genome.
The speed problem comes from having to paste each line separately, storing that in memory and then doing it over and over again for each row, rather than doing them all at once in a vectorised fashion.

Anyway, here is my code so far for that output:

```{rcpp matrices, eval=FALSE}
// Genome output: We have to do this for each trait! So Trait 1 uses mutation types m3 and m4; trait 2 uses m3 and m5, need to get these effects separately


// Trait 1: m3, m4
// Genome output 1 and 2: Binary for whether a mutation has occurred at a given QTL position or not - the segregating sites table, and the size effects of those in the size effects table: The effects of each of these QTL mutations to a separate matrix, the size-effects table
	// Using genomes, subs, and positions from genome output 1 declarations

100000 late() {
	genomes = sim.subpopulations.genomes;
	subs = sim.substitutions[sim.substitutions.mutationType == m3 | sim.substitutions.mutationType == m4];
	positions = sim.getValue("chr_QTL");
	segsitesfile = NULL;
	sizefxfile = NULL;
	
	if (length(subs) == 0 & length(c(genomes.positionsOfMutationsOfType(m3), genomes.positionsOfMutationsOfType(m4))) == 0) {		// Skip the loop if there are no mutations to put in
		segsitesline = rep(0, length(positions));	
		segsitesfile = paste0(c(rep(paste(c(segsitesline, "\n"), ", "), length(genomes) - 1), paste(segsitesline, ", ")));
	// Since if there are no mutations table 1 and 2 will just be 0s and identical, set sizefxfile = segsitesfile
		sizefxfile = segsitesfile;
	}
	
	else {
		
		// For each genome in the seqence, get a list of all the positions where QTL mutations occur, assign a 1 if that position matches a mutation's position 	
		for (genome in genomes) {
		//Table 1: Segregating sites
			segsitesline = rep(0, (sim.chromosome.lastPosition));
			muts = genome.mutations[genome.mutations.mutationType == m3 | genome.mutations.mutationType == m4];
			segsitesline[muts.position] = 1;
			segsitesline[subs.position] = 1;
			segsitesline = segsitesline[positions];
		//Table 2: Size effects
			sizefxline = rep(0.0, sim.chromosome.lastPosition);
			if (muts.size() > 0) { // Double check to make sure we can get those values
			sizefxline[muts.position] = muts.getValue("e0");	// Fill the appropriate vector element in for a given mutation with its selection coef
				}			
			if (subs.size() > 0) {							
			sizefxline[subs.position] = subs.getValue("e0");			
				}
			sizefxline = sizefxline[positions];
			
			if (genome != genomes[genomes.size() - 1]) { // Check if the genome is the last in the list or not: we shouldn't add a new line for the last genome
				segsitesfile = paste0(c(segsitesfile, paste(segsitesline, ", "), "\n"));
				sizefxfile = paste0(c(sizefxfile, paste(sizefxline, ", "), "\n"));
			}
			
			else {
				segsitesfile = paste0(c(segsitesfile, paste(segsitesline, ", ")));
				sizefxfile = paste0(c(sizefxfile, paste(sizefxline, ", ")));
			}
		
		}
	}
	writeFile(asString(paste0(c("~/matrices/model", modelindex, "/", asString(seed), "_", segsitesoutname, "trait1.csv"))), segsitesfile, append = T); // append is true so we can stick outputs from different model/seed runs into the same file
//	// This is ~/model{modelindex}/seed_segsitesoutname_trait1.csv
	writeFile(asString(paste0(c("~/matrices/model", modelindex, "/", asString(seed), "_", sizefxoutname, "trait1.csv"))), sizefxfile, append = T);
	

	
	
	// Genome output 3: per individual homo/heterozygosity or absence: 0, 1, 2 - the genotype table
	// Using subs and positions from genome output 1 declarations
	
	inds = sim.subpopulations.individuals;
	zygosout = NULL;
	zygosfile = NULL;
	
	if (length(subs) == 0 & length(c(inds.genomes.positionsOfMutationsOfType(m3), inds.genomes.positionsOfMutationsOfType(m4))) == 0) {		// Skip the loop if there are no mutations to put in
		zygosline = rep(0, length(positions));	
		zygosfile = paste0(c(rep(paste(c(zygosline, "\n"), ", "), length(inds) - 1), paste(zygosline, ", ")));
	}
	
	else {
		
		// For each individual in the seqence, get a list of all the positions where QTL mutations occur, assign a 1 or 2 if that position matches a mutation's position 	
		for (ind in inds) {
			zygosline = rep(0, (sim.chromosome.lastPosition));
				
			muts0 = ind.genome1.mutations[ind.genome1.mutations.mutationType == m3 | ind.genome1.mutations.mutationType == m4];
			muts1 = ind.genome2.mutations[ind.genome2.mutations.mutationType == m3 | ind.genome2.mutations.mutationType == m4];
			
			// Count the shared mutations (homozygous - appear in both genomes)
			shared = sort(c(setIntersection(muts0, muts1).position, subs.position));
			
			// Then the heterozygous mutations - exist in either genome but not both
			unshared = sort(setSymmetricDifference(muts0, muts1).position);
			
			
			if ( size(shared) > 0 ) {  // For all the mutations that exist in both genomes of the individual, return 2; and 1 for those in only 1 genome
				zygosline[shared] = 2;
			}
			if ( size(unshared) > 0 ) {
				zygosline[unshared] = 1;
			}
		
		zygosline = zygosline[positions];			
			
			if (ind != inds[inds.size() - 1]) { // Check if the individual is the last in the list or not: we shouldn't add a new line for the last individual
				zygosfile = paste0(c(zygosfile, paste(zygosline, ", "), "\n"));
			}
			else {
				zygosfile = paste0(c(zygosfile, paste(zygosline, ", ")));
			}
		
		}
	}
	writeFile(asString(paste0(c("~/matrices/model", modelindex, "/", asString(seed), "_", zygosoutname, "trait1.csv"))), zygosfile, append = T);
	
	

	



}


// Trait 2: m3, m5
// Genome output 1 and 2: Binary for whether a mutation has occurred at a given QTL position or not - the segregating sites table; as well as table 2 for the size effects of those mutations

100000 late() {
	genomes = sim.subpopulations.genomes;
	subs = sim.substitutions[sim.substitutions.mutationType == m3 | sim.substitutions.mutationType == m5];
	positions = sim.getValue("chr_QTL");
	segsitesfile = NULL;
	sizefxfile = NULL;
	
	if (length(subs) == 0 & length(c(genomes.positionsOfMutationsOfType(m3), genomes.positionsOfMutationsOfType(m5))) == 0) {		// Skip the loop if there are no mutations to put in
		segsitesline = rep(0, length(positions));	
		segsitesfile = paste0(c(rep(paste(c(segsitesline, "\n"), ", "), length(genomes) - 1), paste(segsitesline, ", ")));
	// Since if there are no mutations table 1 and 2 will just be 0s and identical, set sizefxfile = segsitesfile
		sizefxfile = segsitesfile;
	}
	
	else {
		
		// For each genome in the seqence, get a list of all the positions where QTL mutations occur, assign a 1 if that position matches a mutation's position 	
		for (genome in genomes) {
		//Table 1: Segregating sites
			segsitesline = rep(0, (sim.chromosome.lastPosition));
			muts = genome.mutations[genome.mutations.mutationType == m3 | genome.mutations.mutationType == m5];
			segsitesline[muts.position] = 1;
			segsitesline[subs.position] = 1;
			segsitesline = segsitesline[positions];
		//Table 2: Size effects
			sizefxline = rep(0.0, sim.chromosome.lastPosition);
			if (muts.size() > 0) { // Double check to make sure we can get those values
			sizefxline[muts.position] = muts.getValue("e1");	// Fill the appropriate vector element in for a given mutation with its selection coef
				}			
			if (subs.size() > 0) {							
			sizefxline[subs.position] = subs.getValue("e1");			
				}
			sizefxline = sizefxline[positions];
			
			if (genome != genomes[genomes.size() - 1]) { // Check if the genome is the last in the list or not: we shouldn't add a new line for the last genome
				segsitesfile = paste0(c(segsitesfile, paste(segsitesline, ", "), "\n"));
				sizefxfile = paste0(c(sizefxfile, paste(sizefxline, ", "), "\n"));
			}
			
			else {
				segsitesfile = paste0(c(segsitesfile, paste(segsitesline, ", ")));
				sizefxfile = paste0(c(sizefxfile, paste(sizefxline, ", ")));
			}
		
		}
	}
	writeFile(asString(paste0(c("~/matrices/model", modelindex, "/", asString(seed), "_", segsitesoutname, "trait2.csv"))), segsitesfile); 

	writeFile(asString(paste0(c("~/matrices/model", modelindex, "/", asString(seed), "_", sizefxoutname, "_trait2.csv"))), sizefxfile);
	
	
	
	// Genome output 3: per individual homo/heterozygosity or absence: 0, 1, 2 - the genotype table
	// Using subs and positions from genome output 1 declarations
	
	inds = sim.subpopulations.individuals;
	zygosout = NULL;
	zygosfile = NULL;
	
	if (length(subs) == 0 & length(c(inds.genomes.positionsOfMutationsOfType(m3), inds.genomes.positionsOfMutationsOfType(m5))) == 0) {		// Skip the loop if there are no mutations to put in
		zygosline = rep(0, length(positions));	
		zygosfile = paste0(c(rep(paste(c(zygosline, "\n"), ", "), length(inds) - 1), paste(zygosline, ", ")));
	}
	
	else {
		
		// For each individual in the seqence, get a list of all the positions where QTL mutations occur, assign a 1 or 2 if that position matches a mutation's position 	
		for (ind in inds) {
			zygosline = rep(0, (sim.chromosome.lastPosition));
				
			muts0 = ind.genome1.mutations[ind.genome1.mutations.mutationType == m3 | ind.genome1.mutations.mutationType == m5];
			muts1 = ind.genome2.mutations[ind.genome2.mutations.mutationType == m3 | ind.genome2.mutations.mutationType == m5];
			
			// Count the shared mutations (homozygous - appear in both genomes)
			shared = sort(c(setIntersection(muts0, muts1).position, subs.position));
			
			// Then the heterozygous mutations - exist in either genome but not both
			unshared = sort(setSymmetricDifference(muts0, muts1).position);
			
			
			if ( size(shared) > 0 ) {  // For all the mutations that exist in both genomes of the individual, return 2; and 1 for those in only 1 genome
				zygosline[shared] = 2;
			}
			if ( size(unshared) > 0 ) {
				zygosline[unshared] = 1;
			}
		
		zygosline = zygosline[positions];			
			
			if (ind != inds[inds.size() - 1]) { // Check if the individual is the last in the list or not: we shouldn't add a new line for the last individual
				zygosfile = paste0(c(zygosfile, paste(zygosline, ", "), "\n"));
			}
			else {
				zygosfile = paste0(c(zygosfile, paste(zygosline, ", ")));
			}
		
		}
	}
	writeFile(asString(paste0(c("~/matrices/model", modelindex, "/", asString(seed), "_", zygosoutname, "trait2.csv"))), zygosfile);
	

}
```

So for each trait there are three tables.
Segregating sites tells if a mutation has occurred at a given site (1) or not (0), and size effects corresponds to that by saying what that mutation's trait value is. Each row is a genome, each column is a QTL locus.
The zygosity table is per individual, and tells how many of the same type of mutation are in each individual - 0, 1 (heterozygous), or 2 (homozygous). Note that two different mutations could be on the same locus and still come back to a value of 2, so it is not true homozygosity, but more how far each individual is from the wild-type.


## Next steps

Next I plan on finishing these systems for four traits (and starting work on 8) and starting work on including balancing selection with a moving optimum. I'll be trying to get runs off to Tinaroo quickly, so Sam, Kevin, Diane, and Daniel can start working on graph theory stuff asap, and also to build my own datasets and knowledge of what parameters are doing in my new models.





